### ISSUE TITLE
InstanceNorm1d throws bad warning

### ISSUE THREAD

**Issue by amithm3** [pytorch/pytorch#115024]
### 🐛 Describe the bug

nn.InstanceNorm1d is throwing `UserWarning: input's size at dim=0 does not match num_features. You can silence this warning by not passing in num_features, which is not used because affine=False
  warnings.warn(f"input's size at dim={feature_dim} does not match num_features. "` even though everything is correct
  
```py
import torch
from torch import nn

norm = nn.InstanceNorm1d(20)  # features
out = norm(torch.randn(32, 20)) # (batch_size, features)
out.shape # (32, 20)
``` 

### Versions

PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.5 (v3.11.5:cce6ba91b3, Aug 24 2023, 10:50:31) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-14.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] numpy==1.26.1
[pip3] torch==2.1.0
[pip3] torchvision==0.16.0
[conda] Could not collect

**Comment by drisspg** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.InstanceNorm1d.html#torch.nn.InstanceNorm1d

the for 2d inputs the channels comes first. I think that if you want batched input it needs to be 3d.

**Comment by phrasenmaeher** [comment]
In the case that one still is working with, say, timeseries data, how could we then use InstanceNorm1d with data of shape `(batch_size, features)`?
E.g., this raises the warning:

```python
import torch
from torch import nn

norm = nn.InstanceNorm1d(20)  # features
out = norm(torch.randn(32, 20)) # (batch_size, features)
out.shape # (32, 20)
```

while switching the dimensions does not:
```python
import torch
from torch import nn

norm = nn.InstanceNorm1d(20)  # features
out = norm(torch.randn(20, 32)) # (batch_size, features)
out.shape # (32, 20)
```
I think it would be too costly to dynamically switch features and batch_size during training.
Is there a better way?
